watsonx/runtime/src/main/java/io/quarkiverse/langchain4j/watsonx/client/WatsonRestApi.java

```diff
@@ -0,0 +1,198 @@
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.ProcessingException;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.reactive.ClientWebApplicationException;
+import org.jboss.resteasy.reactive.client.api.ClientLogger;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.quarkiverse.langchain4j.QuarkusJsonCodecFactory;
+import io.quarkiverse.langchain4j.watsonx.bean.TextGenerationRequest;
+import io.quarkiverse.langchain4j.watsonx.bean.TextGenerationResponse;
+import io.quarkiverse.langchain4j.watsonx.bean.WatsonError;
+import io.quarkiverse.langchain4j.watsonx.client.filter.BearerRequestFilter;
+import io.quarkiverse.langchain4j.watsonx.exception.WatsonException;
+import io.quarkus.rest.client.reactive.ClientExceptionMapper;
+import io.quarkus.rest.client.reactive.jackson.ClientObjectMapper;
+import io.vertx.core.Handler;
+import io.vertx.core.MultiMap;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpClientResponse;
+
+/**
+ * This Microprofile REST client is used as the building block of all the API
+ * calls to Watsonx. The implementation is provided by the Reactive REST Client
+ * in Quarkus.
+ */
+@Path("/ml/v1-beta")
+@RegisterProvider(BearerRequestFilter.class)
+@Consumes(MediaType.APPLICATION_JSON)
+@Produces(MediaType.APPLICATION_JSON)
+public interface WatsonRestApi {
+
+    static Logger logger = Logger.getLogger(WatsonRestApi.class);
+
+    @POST
+    @Path("generation/text")
+    TextGenerationResponse chat(TextGenerationRequest request, @QueryParam("version") String version) throws WatsonException;
+
+    @ClientExceptionMapper
+    static WatsonException toException(jakarta.ws.rs.core.Response response) {
+
+        if (MediaType.TEXT_PLAIN.equals(response.getHeaderString("Content-Type")))
+            return new WatsonException(response.readEntity(String.class), response.getStatus());
+
+        try {
+
+            WatsonError ex = response.readEntity(WatsonError.class);
+
+            StringJoiner joiner = new StringJoiner("\n");
+            if (ex.errors() != null && ex.errors().size() > 0) {
+                for (WatsonError.Error error : ex.errors())
+                    joiner.add("%s: %s".formatted(error.code(), error.message()));
+            }
+
+            return new WatsonException(joiner.toString(), response.getStatus(), ex);
+
+        } catch (ClientWebApplicationException | ProcessingException e) {
+            return new WatsonException(e.getCause(), response.getStatus());
+        }
+    }
+
+    @ClientObjectMapper
+    static ObjectMapper objectMapper(ObjectMapper defaultObjectMapper) {
+        return QuarkusJsonCodecFactory.SnakeCaseObjectMapperHolder.MAPPER;
```

Comment:
Does WatsonX really use snake case in it's JSON? If not, this should probably use camel case or whatever :)

Reply:
Yes, this is an example of response from watsonx.
```
{
    "model_id": "meta-llama/llama-2-70b-chat",
    "created_at": "2024-01-23T13:12:46.370Z",
    "results": [
        {
            "generated_text": "...",
            "generated_token_count": 9,
            "input_token_count": 50,
            "stop_reason": "eos_token",
            "seed": 588258570
        }
    ],
 ...
 ```